server/src/main/java/io/crate/expression/predicate/NotPredicate.java

```diff
@@ -142,27 +140,28 @@ public Void visitFunction(Function function, NullabilityContext context) {
                 var b = function.arguments().get(1);
                 if (a instanceof Reference ref && b instanceof Literal<?>) {
                     if (ref.valueType().id() == DataTypes.UNTYPED_OBJECT.id()) {
-                        context.removeNullValues = true;
                         return null;
                     }
                 }
             } else if (Ignore3vlFunction.NAME.equals(functionName)) {
-                context.useNotQuery = true;
+                context.isNullable = false;
                 return null;
             } else {
                 var signature = function.signature();
-                if (signature.hasFeature(Feature.NULLABLE)) {
-                    context.removeNullValues = true;
-                } else if (signature.hasFeature(Feature.NON_NULLABLE)) {
-                    context.useNotQuery = true;
-                } else {
+                if (signature.hasFeature(Feature.NON_NULLABLE)) {
+                    context.isNullable = false;
+                } else if (!signature.hasFeature(Feature.NULLABLE)) {
                     // default case
                     context.enforceThreeValuedLogic = true;
                     return null;
                 }
             }
+            // saves and restores isNullable of the current context
+            // such that any non-nullables observed from the left arg is not transferred to the right arg.
+            boolean isNullable = context.isNullable;
             for (Symbol arg : function.arguments()) {
                 arg.accept(this, context);
+                context.isNullable = isNullable;
```

Comment:
Thank you for reviewing @mkleen . I think the static method `of` was the cause of the confusion. Please check again.